src/build.rs

```diff
@@ -0,0 +1,250 @@
+}
+
+pub fn output_link_arg(arg: impl AsRef<str>) {
+    println!("cargo:rustc-link-arg={}", arg.as_ref());
+}
+
+#[derive(Clone, Debug)]
+pub struct CInclArgs(String);
+
+impl TryFrom<&SconsVariables> for CInclArgs {
+    type Error = anyhow::Error;
+
+    fn try_from(scons: &SconsVariables) -> Result<Self> {
+        Ok(Self(scons.incflags.clone()))
+    }
+}
+
+impl CInclArgs {
+    pub fn propagate(&self) {
+        output(VAR_C_INCLUDE_ARGS, self.0.as_str());
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct LinkArgsBuilder {
+    pub libflags: Vec<String>,
+    pub linkflags: Vec<String>,
+    pub libdirflags: Vec<String>,
+    pub(crate) use_linkproxy: bool,
+    /// The path to the linker executable.
+    pub(crate) linker: Option<PathBuf>,
+    /// The working directory that should be set when linking.
+    pub(crate) working_directory: Option<PathBuf>,
+    pub(crate) dedup_libs: bool,
+}
+
+impl LinkArgsBuilder {
+    pub fn use_linkproxy(&mut self, value: bool) -> &mut Self {
+        self.use_linkproxy = value;
+        self
+    }
+
+    pub fn working_directory(&mut self, dir: impl AsRef<Path>) -> &mut Self {
+        self.working_directory = Some(dir.as_ref().to_owned());
+        self
+    }
+
+    pub fn dedup_libs(&mut self, dedup: bool) -> &mut Self {
+        self.dedup_libs = dedup;
+        self
+    }
+
+    pub fn build(&mut self, project_path: impl AsRef<Path>) -> LinkArgs {
+        let mut result = Vec::new();
+
+        if self.use_linkproxy {
+            if let Some(linker) = &self.linker {
+                result.push(format!(
+                    "{}{}",
+                    LINKPROXY_LINKER_ARG,
+                    linker.display()
+                ));
+            }
+
+            if self.dedup_libs {
+                result.push(LINKPROXY_DEDUP_LIBS_ARG.into());
+            }
+        }
+
+        // A hack to workaround this issue with Rust's compiler intrinsics: https://github.com/rust-lang/compiler-builtins/issues/353
+        //result.push("-Wl,--allow-multiple-definition".to_owned());
+
+        result.push(format!("-L{}", project_path.as_ref().display()));
+
+        for arg in &self.libdirflags {
+            result.push(arg.clone());
+        }
+
+        for arg in &self.libflags {
+            // Hack: convert the relative paths that Pio generates to absolute ones
+            let arg = if arg.starts_with(".pio/") {
+                format!("{}/{}", project_path.as_ref().display(), arg)
+            } else if arg.starts_with(".pio\\") {
+                format!("{}\\{}", project_path.as_ref().display(), arg)
+            } else {
+                arg.clone()
+            };
+
+            result.push(arg);
+        }
+
+        for arg in &self.linkflags {
+            result.push(arg.clone());
+        }
+
+        LinkArgs { args: result }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct LinkArgs {
+    args: Vec<String>,
+}
+
+impl LinkArgs {
```

Comment:
Future, not for now, and fyi: there is now a common pattern shared between the "push linker args to the binary crate" (or any interested crate), "push kconfig Rustc cfg flags to any interested crate" and "push include directories to any interested crate" (not that this last one is used by anybody). It is the mechanism where we are encoding all of this data into an environment variable, and later on, in the consuming crate, we are decoding this data from the environment variable and output-ing it with `println!("cargo:...)` statements.

Ideally we should be seeking to abstract this functionality into its own utility, however this can wait for after these changes are implemented.